### PR TITLE
Use new UEFI bootloader

### PR DESCRIPTION
**Pull request by tsoutsman**
The bootloader is at <https://github.com/theseus-os/uefi-bootloader>. It will also need to be reviewed.

Depends on #785, #786, and #790. I haven't been able to properly review all the changes in this PR because it builds on top of many different PRs. Once they are merged, I'll be able to review the changes properly.

### CODE REVIEW

File: kernel/nano_core/src/asm/defines.asm [hunk 1]
```diff
@@ -16,6 +16,7 @@ AP_NMI_LINT         equ TRAMPOLINE + 56
 AP_NMI_FLAGS        equ TRAMPOLINE + 64
 AP_MAX_FB_WIDTH     equ TRAMPOLINE + 72
 AP_MAX_FB_HEIGHT    equ TRAMPOLINE + 80
+AP_GDT				equ TRAMPOLINE + 88
```
Comment (tsoutsman) [inline kernel/nano_core/src/asm/defines.asm]: This shows up differently in my editor.

**Comment by kevinaboos** [comment]
Thanks, I've reviewed all the PR this depends on and am ready to merge them pending some quick discussions.

File: kernel/nano_core/src/asm/defines.asm (hunk 1 above)
Comment (kevinaboos) [reply kernel/nano_core/src/asm/defines.asm]: looks like you've used tabs instead of spaces

**Comment by tsoutsman** [comment]
Depends on #803.

File: kernel/boot_info/src/lib.rs [hunk 2]
```diff
@@ -122,7 +122,7 @@ pub trait BootInformation: 'static {
     ) -> Result<Self::AdditionalReservedMemoryRegions, &'static str>;
 
     /// Returns the end of the kernel's image in physical memory.
-    fn kernel_end(&self) -> Result<PhysicalAddress, &'static str>;
+    fn kernel_end(&self) -> Result<VirtualAddress, &'static str>;
```
Comment (kevinaboos) [inline kernel/boot_info/src/lib.rs:126]: fix doc comment above

File: kernel/boot_info/src/lib.rs [hunk 3]
```diff
@@ -38,7 +38,7 @@ pub trait ElfSection {
     /// Returns the section's starting virtual address.
     fn start(&self) -> VirtualAddress;
 
-    /// Returns the section's length.
+    /// Returns the section's length in memory.
```
Comment (kevinaboos) [inline kernel/boot_info/src/lib.rs]: is this comment addition intended to contrast against the section's length in the ELF file? if so, let's say that too.

File: kernel/nano_core/src/lib.rs [hunk 4]
```diff
@@ -194,10 +214,10 @@ where
             .ok_or("no single symbol matching \"captain::init\"")?;
         log::info!("The nano_core (in loadable mode) is invoking the captain init function: {:?}", section.name);
 
-        type CaptainInitFunc = fn(MmiRef, NoDrop<EarlyIdentityMappedPages>, NoDrop<Stack>, VirtualAddress, VirtualAddress, Option<PhysicalAddress>) -> Result<(), &'static str>;
+        type CaptainInitFunc = fn(MmiRef, NoDrop<EarlyIdentityMappedPages>, NoDrop<Stack>, VirtualAddress, VirtualAddress, VirtualAddress, Option<PhysicalAddress>) -> Result<(), &'static str>;
```
Comment (kevinaboos) [inline kernel/nano_core/src/lib.rs:217]: according to your other docs, the ap_gdt is a physical address

File: kernel/memory/src/lib.rs [hunk 5]
```diff
@@ -220,7 +221,15 @@ pub fn init(
     debug!("Initialized new frame allocator!");
     frame_allocator::dump_frame_allocator_state();
 
-    page_allocator::init(VirtualAddress::new_canonical(boot_info.kernel_end()?.value()))?;
+    page_allocator::init(
+        VirtualAddress::new(
+            Mapper::from_current()
+                .translate(boot_info.kernel_end()?)
```
Comment (kevinaboos) [inline kernel/memory/src/lib.rs]: is this right? Looking at the calculation of `kernel_end`, it seems like an exclusive upper bound. I didn't dig into it in detail, but what happens if `kernel_end()` is aligned to the start of the next page, causing this `translate()` action to fail?

Comment (kevinaboos) [reply kernel/memory/src/lib.rs]: ~~ok on second inspection, this is a bit odd. The boot_info provides the kernel end vaddr, which we then translate to physical, which we the use to obtain the identity vaddr? I legit cannot tell if that's equivalent to the previous code, lol~~
EDIT: nvm, i'm tired lol. The answer is clearly that the boot loader isn't guaranteed to have identity mapped it.

File: tools/uefi_builder/common/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "uefi_builder_common"
+version = "0.1.0"
+edition = "2021"
+authors = ["Klim Tsoutsman <klim@tsoutsman.com>"]
```
Comment (kevinaboos) [inline tools/uefi_builder/common/Cargo.toml:4]: please add a description (and in other Cargo.toml files too)

**Review by kevinaboos — COMMENTED** [review]
Minor to-do: add a description of the `uefi_builder*` tool crates to the `tools/README.md` file

File: kernel/memory/src/paging/mapper.rs [hunk 7]
```diff
@@ -50,6 +50,9 @@ use owned_borrowed_trait::{OwnedOrBorrowed, Owned, Borrowed};
 /// that it is only invoked for `UnmappedFrames`.
 pub(super) static INTO_ALLOCATED_FRAMES_FUNC: Once<fn(FrameRange) -> AllocatedFrames> = Once::new();
 
+pub fn translate(virtual_address: VirtualAddress) -> Option<PhysicalAddress> {
```
Comment (kevinaboos) [inline kernel/memory/src/paging/mapper.rs:55]:
```suggestion
/// A convenience function to translate the given virtual address into a
/// physical address using the currently-active page table.
pub fn translate(virtual_address: VirtualAddress) -> Option<PhysicalAddress> {
```

**Review by kevinaboos — CHANGES_REQUESTED** [review]

**Comment by kevinaboos** [comment]
LGTM; will approve & merge pending a review of `uefi-bootloader`

**Review by kevinaboos — APPROVED** [review]

**Pull request merged**